### PR TITLE
Ильин Максим - Лабораторная работа №2

### PR DESCRIPTION
**Pull request by Hetzeer**

### CODE REVIEW

**Comment by AlexIK-0** [comment]
All good

**Comment by MatveevSE1** [comment]
Nice lab

File: modules/bisymmetric_matrix/include/bisymmetric_matrix.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include <vector>
+
+using std::vector;
+
+class BisymmetricMatrix {
+ private:
+  vector<vector<double>> container;
+  int size;
+
+  bool isSymmetric(vector<vector<double>> mat);
```
Comment (dkurt) [inline modules/bisymmetric_matrix/include/bisymmetric_matrix.h:15]: Memory overhead. Use constant reference

File: modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+#include <random>
+#include <vector>
+
+// using std::vector;
+
+bool BisymmetricMatrix::isSymmetric(vector<vector<double>> mat) {
+  int N = static_cast<int>(mat.size());
+
+  vector<vector<double>> temp = transpose(mat);
```
Comment (dkurt) [inline modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp:15]: This method can be implemented without any transpose

File: modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+  for (int i = 0; i < N; i++)
+    for (int j = 0; j < N; j++)
+      if (mat[i][j] != temp[i][j]) return false;
+  return true;
+}
+
+BisymmetricMatrix::BisymmetricMatrix() {
+  this->container = vector<vector<double>>();
```
Comment (dkurt) [inline modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp:24]: not needed in default constructor

File: modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+BisymmetricMatrix::BisymmetricMatrix(const BisymmetricMatrix& other) {
+  this->container = other.container;
+  this->size = other.size;
+}
+
+BisymmetricMatrix::BisymmetricMatrix(const vector<vector<double>>& other) {
+  if (!isSymmetric(other)) {
+    throw -1;
```
Comment (dkurt) [inline modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp:35]: throw a meaningful exception:
```cpp
throw std::exception("some messsage")
```

File: modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  this->container = other.container;
+  this->size = other.size;
+  return *this;
+}
+
+BisymmetricMatrix& BisymmetricMatrix::operator+(
+    const BisymmetricMatrix& other) {
+  // if (this->size != other.size) {
+  //   throw -1;
+  // }
```
Comment (dkurt) [inline modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp:60]: Remove commented code

File: modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+  for (int i = 0; i < size; i++) {
+    for (int j = i; j < size; j++) {
+      (*this)[i][j] = (*this)[i][j] + other[i][j];
+      (*this)[j][i] = (*this)[i][j];
```
Comment (dkurt) [inline modules/bisymmetric_matrix/src/bisymmetric_matrix.cpp:65]: This is ineffective. Use `container`

**Pull request closed without merging**